pkg/hostpath/controllerserver.go

```diff
@@ -139,3 +145,227 @@ func (cs *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req
 	}
 	return &csi.ValidateVolumeCapabilitiesResponse{Supported: true, Message: ""}, nil
 }
+
+// CreateSnapshot uses tar command to create snapshot for hostpath volume. The tar command can quickly create
+// archives of entire directories. The host image must have "tar" binaries in /bin, /usr/sbin, or /usr/bin.
+func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		glog.V(3).Infof("invalid create snapshot req: %v", req)
+		return nil, err
+	}
+
+	if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Name missing in request")
+	}
+	// Check arguments
+	if len(req.GetSourceVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "SourceVolumeId missing in request")
+	}
+
+	// Need to check for already existing snapshot name, and if found check for the
+	// requested sourceVolumeId and sourceVolumeId of snapshot that has been created.
+	if exSnap, err := getSnapshotByName(req.GetName()); err == nil {
+		// Since err is nil, it means the snapshot with the same name already exists need
+		// to check if the sourceVolumeId of existing snapshot is the same as in new request.
+		if exSnap.VolID == req.GetSourceVolumeId() {
+			// same snapshot has been created.
+			return &csi.CreateSnapshotResponse{
+				Snapshot: &csi.Snapshot{
+					Id:             exSnap.Id,
+					SourceVolumeId: exSnap.VolID,
+					CreatedAt:      exSnap.CreateAt,
+					SizeBytes:      exSnap.SizeBytes,
+					Status:         exSnap.Status,
+				},
+			}, nil
+		}
+		return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("snapshot with the same name: %s but with different SourceVolumeId already exist", req.GetName()))
+	}
+
+	volumeID := req.GetSourceVolumeId()
+	hostPathVolume, ok := hostPathVolumes[volumeID]
+	if !ok {
+		return nil, status.Error(codes.Internal, "volumeID is not exist")
+	}
+
+	snapshotID := uuid.NewUUID().String()
+	createAt := time.Now().UnixNano()
+	volPath := hostPathVolume.VolPath
+	file := snapshotRoot + snapshotID + ".tgz"
+	args := []string{"czf", file, "-C", volPath, "."}
+	executor := utilexec.New()
+	out, err := executor.Command("tar", args...).CombinedOutput()
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed create snapshot: %v: %s", err, out))
+	}
+
+	fileInfo, err := os.Stat(file)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed create snapshot: %v", err))
+	}
+
+	glog.V(4).Infof("create volume snapshot %s", file)
+	snapshot := hostPathSnapshot{}
+	snapshot.Name = req.GetName()
+	snapshot.Id = snapshotID
+	snapshot.VolID = volumeID
+	snapshot.Path = file
+	snapshot.CreateAt = createAt
+	snapshot.SizeBytes = fileInfo.Size()
```

Comment:
This should be the size of the original volume, not how much space is used.
For example, "hostPathVol.VolSize = capacity" in CreateVolume.  We should do the same thing here.

Reply:
Thanks for clarify this, Updated